src/ark/utils/data_utils.py

```diff
@@ -54,119 +57,199 @@ def save_fov_mask(fov, data_dir, mask_data, sub_dir=None, name_suffix=''):
 
 
-def relabel_segmentation(labeled_image, labels_dict):
-    """Takes a labeled image and translates its labels according to a dictionary.
-
-    Returns the relabeled array (according to the dictionary).
-
-    Args:
-        labeled_image (numpy.ndarray):
-            2D numpy array of labeled cell objects.
-        labels_dict (dict):
-            a mapping between labeled cells and their clusters.
-
-    Returns:
-        numpy.ndarray:
-            The relabeled array.
+@dataclass
+class CellClusterMaskData:
     """
+    A dataclass containing the cell data, cell label column, cluster column and the mapping from a
+    cell label to a cluster.
 
-    img = np.copy(labeled_image)
-    unique_cell_ids = np.unique(labeled_image)
-    unique_cell_ids = unique_cell_ids[np.nonzero(unique_cell_ids)]
-
-    default_label = max(labels_dict.values()) + 1
-
-    # cast to int16 to allow for Photoshop loading
-    relabeled_img = np.vectorize(
-        lambda x: labels_dict.get(x, default_label) if x != 0 else 0
-    )(img).astype('int16')
-
-    return relabeled_img
+    Args:
+        data (pd.DataFrame):
+            A cell table with the cell label column and the cluster column.
+        fov_col (str):
+            The name of the column in the cell table that contains the FOV ID.
+        label_column (str):
+            The name of the column in the cell table that contains the cell label.
+        cluster_column (str):
+            The name of the column in the cell table that contains the cluster label.
 
 
-def label_cells_by_cluster(fov, all_data, label_map, fov_col=settings.FOV_ID,
-                           cell_label_column=settings.CELL_LABEL,
-                           cluster_column=settings.KMEANS_CLUSTER):
-    """Translates cell-ID labeled images according to the clustering assignment.
+    Fields:
+        unique_fovs (List[str]):
+            A list of all unique FOVs in the cell table.
+        unassigned_id (int):
+            The cluster ID that is assigned to cells that are not assigned to a cluster.
+        mapping (pd.core.groupby.DataFrameGroupBy):
+            A GroupBy object that contains the mapping from the cell label to the cluster label for
+            each FOV, sorted by Segmentation label.
+    """
+    data: pd.DataFrame
+    fov_col: str
+    label_column: str
+    cluster_column: str
+
+    # Initialized in __post_init__
+    unique_fovs: List[str] = field(init=False)
+    unassigned_id: int = field(init=False)
+    mapping: DataFrameGroupBy = field(init=False)
+
+    def __post_init__(self) -> None:
+
+        self.data[self.label_column] = self.data[self.label_column].astype(np.int64)
+        self.data[self.cluster_column] = self.data[self.cluster_column].astype(np.int64)
+
+        self.unique_fovs = ns.natsorted(self.data[self.fov_col].unique().tolist())
+
+        self.unassigned_id: int = self.data[self.cluster_column].max() + 1
+
+        self.mapping: DataFrameGroupBy = (
+            pd.concat(
```

Comment:
This is confusing to read, especially without comments. I would think about splitting this up into multiple actions, rather than a single "one liner" that spans multiple lines